sklearn/base.py

```diff
@@ -119,6 +119,12 @@ class BaseEstimator:
     arguments (no ``*args`` or ``**kwargs``).
     """
 
+    def __new__(cls, *args, **kwargs):
+        est = super().__new__(cls)
+        if cls.__module__.startswith("sklearn."):
+            est.__sklearn_version__ = __version__
```

Comment:
I think we should not introduced our own "dunder" attributes. Just using a private a attribute should be enough:

```suggestion
            est._sklearn_version = __version__
```

Reply:
Note that the rest of the code would need to be adapted accordingly when accepting this suggestion.

Reply:
For reference:

https://peps.python.org/pep-0008/#descriptive-naming-styles

> `__double_leading_and_trailing_underscore__`: “magic” objects or attributes that live in user-controlled namespaces. E.g. `__init__`, `__import__` or `__file__`. Never invent such names; only use them as documented.

Reply:
I think we're already down the path of "dunder is how we signal dev API but not user API". We already have `__sklearn_is_fitted__`, and we just approved the `__sklearn_clone__` SLEP.

Here, the question is: do we want to expose this attribute in a way which people should rely on and read, or is it only a private thing for internal use. Since internally, we don't actually need it, and the issue started by a user/dev requesting to have access to this attribute, a private attribute with no guarantees on backward compatibility doesn't make much sense. We can either expose it as a public attribute without dunder, or with.

And we already use `__sklearn_is_fitted__` inside sklearn, and we might do that with `__sklearn_clone__` as well. Therefore to me it is sensible to have a `__sklearn_version__` here.

Reply:
`__protocol__` is common pattern with Python numerical libraries for creating protocols. For example:

- `NumPy` has `__array_function__`, `__array_interface__`, `__array__`, etc.
- DataFrame exchange protocol has `__dataframe__`
- Dask has `__dask_graph__`
- CUDA GPU array libraries use `__cuda_array_interface__`